### ISSUE TITLE
Proficiencies, crafting, failure chance, UI, etc

### ISSUE THREAD

**Issue by Kelenius** [CleverRaven/Cataclysm-DDA#48981]
# Describe the bug

I would like to collect all issues and RPs that are currently opened [(there are too many, so my solution for solving that is to open one more)](https://xkcd.com/927/), because there is a significant overlap between them and they boil down to the same problem:

**The effect of having missing proficiencies on crafting and crafting failures in general are broken in various ways.**

# Existing discussions

Issues:
#46662
#47566
#48465
#48597
#48925
#48956

PRs:
#48658
#46082
#46153
#48673
#48650
#48924

And a discussion:
#48912

# Why I'm opening this

Because all of this is very closely related and it's essentially the same problem that should be resolved in one go, not broken up into a bunch of PRs - merging #48673 fixed one problem, but also made crafting even very simple items practically impossible without proficiencies _and_ left missing proficiencies impossible to mitigate because that's also broken - a worse situation, in my opinion!

I'd like to bring all discussion into the same spot, because right now it's spread out and is difficult to follow.

# Let's untangle the knot

Here's the full list of what's currently bugged:

- Any books that mitigate proficiency penalties are only applied to the UI. This is happening because for some reason - something about performance? - there are entirely separate functions for calculating time and failure chances of a recipe for the UI and for the actual crafting: UI's are in `recipe.cpp`, `proficiency_time_maluses` and `proficiency_failure_maluses`, while crafting's failures are handled in `crafting.cpp`, `crafting_success_roll`, and time in `recipe.cpp`, `time_to_craft_moves`.
- Books that mitigate proficiency penalties also don't do it correctly. They simply multiply failure/time by their mitigation modifier, which leads to `mag_animecon` Boston AnimeCon magazine having 0.1, and you have `prof_leatherworking_basic`'s time multiplier of 1.5 being multiplied by 0.1, resulting in 0.15 crafting time multiplier overall - or about 6 times faster. HOWEVER! Due to the bug above, this is only applied to the UI. In reality, the books have no effects on crafting time or failure chances. Relevant functions are, again, `recipe.cpp`, `proficiency_time_maluses` and `proficiency_failure_maluses`.
- To make this even more confusing, there is a third place where those penalties are calculated: `recipe.cpp`, `missing_proficiencies_string`, which shows that "Proficiencies Missing:" line. It also simply multiplies time/failure chance by book's factor, but it also caps it at a minimum of 1.
- But there's more! That "Time to complete:" line? Uses, ultimately, `recipe.cpp`, `time_to_craft_moves`, which means it's accurate to the actual crafting time, and isn't affected by the books.
- Failure rate is handled very poorly. Relevant code is in `activity_actor.cpp`, `craft_activity_actor::do_turn`, near the end, going to `crafting.cpp`, `item::handle_craft_failure`, and then `Character::crafting_success_roll`. Here's the thing: your failures are checked at specific checkpoints; the next checkpoint is set when you start crafting and whenever you trigger one, and it's placed between your current progress and 100%, depending on the failure rate; being at 50% progress and rolling 0.1 would place the next check at 55%. Essentially, failure rate is double-dipping: it's both making the checks more frequent and making the actual failure chance higher. For example, having failure rate at static 0.5 (not actually possible, but bear with me) would put a checkpoint at 50%, then 75%, then 87.5%, and so on, and then on each of those you'd have a 50% chance of destroying every item in the craft - and every time you will also lose between 0.5 * 25 and 0.5 * 35 percent of the progress. And every time you are set back, the checkpoints are moved again - if you fail at 50% and get set back by 10%, you're now at 40% and your next checkpoint is at 70%. This creates a huge problem as your approach 100%, because the checks become more and more frequent, and you get set back by the same flat amount, which makes it insanely hard to reach 100% even with fairly low success rate. And to make it worse...
- The impact the missing proficiencies have is absolutely insane. The actual check is, of course, not a flat 0.5. It's an opposed roll between your skill (number dice equal to weighted average between the primary and secondary skill, with 16+int dice sides) and craft difficulty (number dice equal to weighted average between the primary and secondary skill, with 24 dice sides); the skill roll is divided by the difficulty roll, and the result is your success rate. It is also divided by the effect of any missing proficiencies, which tends to absolutely destroy any success chance, especially when combined with the previous problem. For example, Principles of Leatherworking has a 5x failure rate.
- A relevant problem is that a recipe that uses many different components is more susceptible to failures - every time a failure is checked, it rolls against every component.

# How to resolve this mess

First we need to stop and asks ourselves: ~~why am I spending my Friday evening digging through C++ code~~ how difficult should it be for different characters to craft? There are three factors at play:

- Character's intelligence vs the normal average of 8.
- Character's skills (primary and secondaries, if any) vs craft difficulty
- Any missing proficiencies that affect failure rate and whether or no they are mitigated; and if they are, then to which degree

(there are also NPCs and other small factors like crafting in darkness, but let's put that aside)

I see a lot of people arguing about how math works. That is not the approach that we should take. It would be, in my opinion, best to answer these:

"A character with 8 intelligence and the skill matching craft's difficulty should succeed XX% of the time."
"If that character is missing a 2x failure chance, they should succeed YY% of the time, which is halved success rate / doubled failure rate / something else, according to a formula."
"If that character instead has Z or more intelligence, they should always succeed."
"If that character's skill is instead W or more % higher than craft's difficulty, they should always succeed."
"If that character has 4 intelligence, their success rate should be (I am running out of good letters to use)%."
"If that character has skill (whatever)% lower than craft's difficulty, their success rate should be (something)%."

And then adjust the results to fit these. Failure checks should be more evenly spread across the recipe, and failures themselves should be more manageable. Right now they send you into a death spiral where you can't finish the craft because you keep getting set back by failures.

Second, and this is important, there needs to be one unified function for getting the penalties. Right now it's in four different places: `recipe.cpp`, `proficiency_time_maluses` and `proficiency_failure_maluses`; `crafting.cpp`, `crafting_success_roll`; `recipe.cpp`, `time_to_craft_moves`; `recipe.cpp`, `missing_proficiencies_string`.

Third, there needs to be a clear UI that shows what your actual failure chance is and how the proficiencies missing are affecting it. "Missing this proficiency makes you twice as likely to fail" isn't useful information when you don't know the base failure rate and how that 2x is applied to it. Showing how much proficiency progression you'd get from crafting the item is also something to consider.

**Comment by actual-nh** [comment]
Fourth, the code needs to be better-documented so it's clear to see exactly what it happening...

**Comment by deadcutlass2** [comment]
> * your failures are checked at specific checkpoints. These checkpoints occur more frequently when your failure rate is high...

They are not. The output of Character::crafting_success_roll sets one singular failure point for the craft based on it's output. This is done in item::set_next_failure_point, line 1054 of crafting.cpp.

Attempting the same craft from scratch and attempting it at 99% have the same odds of succeeding. The difference is, assuming you roll a 0.6 from Character::crafting_success_roll, you'd end up at 60% completion and 99.6% completion respectively before failing.

#48925 and #48956 should be resolved by #48924
#46662 and #48465 should be resolved by #48650 and #48658
#48597 should be resolved by #46082

#46153 should be the long term resolution to all of this.

**Comment by Kelenius** [comment]
> They are not. The output of Character::crafting_success_roll sets one singular failure point for the craft based on it's output. This is done in item::set_next_failure_point, line 1054 of crafting.cpp.

It's called multiple times, setting multiple failure points.

**Comment by pehamm** [comment]
That is semantics, only one failure point exists at any one time, but the higher your expected failure, the higher the expected total number of failures until completion.

**Comment by Kelenius** [comment]
> That is semantics, only one failure point exists at any one time, but the higher your expected failure, the higher the expected total number of failures until completion.

I don't think you understand. Every time you trigger a failure, the next failure point is reset. In your example, if you are starting with 0% and 0.2 success rate, your next failure is 20% away, and if you start with 99%, it's 0.2% away. This creates a "clustering" effect where failures become more frequent as your approach 100%. And since every failure sets back your progress, that makes it very difficult to finish the craft.

**Comment by deadcutlass2** [comment]
your phrasing indicated that the failure points are pre-chosen before even attempting the recipe, and that they were pre-chosen at the high end of a recipe. They are generated on the fly was my point.

**Comment by Kelenius** [comment]
I suppose that's worth clarifying, but doesn't change my overall point.

**Comment by deadcutlass2** [comment]
Are we waiting on I-am-Erk to check through the PRs or is Kevin willing to do them?

**Comment by Tommy-os** [comment]
I'm concerned that the "0.F Feature Freeze" tag means the soon(tm) stable will include the current 'impossible' crafting.

**Comment by actual-nh** [comment]
> I'm concerned that the "0.F Feature Freeze" tag means the soon(tm) stable will include the current 'impossible' crafting.

I can remove it if you'd like; it was more to make sure this doesn't go stale, and that I suggest that _further_ fixing should _definitely_ be a 0.G blocker.

**Comment by Evrze** [comment]
That is because the probability is tied to the item itself with the progress, if you have a 20% chance to pass each check you are expected to have an average pass rate of 20 out of 100 checks it presents to you. If it detects that you are passing too many checks and succeeding too often in the beginning, it starts to make you fail much more to compensate. The probability is reflected in the frequency of events to preserve the meaning of "20% chance to pass" you can treat this as a pool.

If you succeed too much you have depleted your average amount of success pool and need to start filling a failure quota, consecutive successes lower your amount of success chance by nature (or rather, have a lower frequency).

The more you succeed, the lower your effective success for passing threshold (to preserve what the pass rate means), because you have completed your success quota and it's time to fail, the more you fail the more the success chance can be "regenerated" and then have a chance to succeed.

With a high pass rate, your success will be lowered the more times you succeed, especially consecutively, and it will generate the failure threshold points depending on the expected result of failing the average amount and reflect it in a frequency of an event, this leads to the so-called clustering of failures, if you depleted your pool of expected successes, you will start to be forced to fail to fulfill the expected amount of failures the end, this regenerates the pool for success, then you can potentially finish the item.

The pool is tied to the item craft itself trying to preserve the frequency of events. At some point, you have to start "failure crafting" because it's necessary for the item to accept the completion of item progress. 

Succeeding too much early can be setting yourself up for failure. 

Alternatively, failing early is a form of "failstacking" for success later, adding a bonus to effective success (compared to the last check) and are a necessary component to work towards item completion.

**Comment by pehamm** [comment]
> That is because the probability is tied to the item itself with the progress, if you have a 20% chance to pass each check you are expected to have an average pass rate of 20 out of 100 checks it presents to you. If it detects that you are passing too many checks and succeeding too often in the beginning, it starts to make you fail much more to compensate. The probability is reflected in the frequency of events to preserve the meaning of "20% chance to pass" you can treat this as a pool.

This is not how any of this works. It does not matter if you failed 10 times or none and then fail again at 50%, your probability of more failures going forward with the same craft is exactly the same. Look up Gambler's Fallacy

**Comment by pehamm** [comment]
I made some plots to illustrate how the proficiency factor changes the success probability. This is a recipe with 5 difficulty and attempted with 5 skill, 8 intelligence. No proficiency missing. To succeed the result of `skill_roll` has to be larger than `diff_roll`, or `skill / diff_roll > 1`. The left plot shows the density function of both rolls, which lie on top of each other because we fulfill the requirements exactly. The second plot shows the density function of the ratio `skill_roll / diff_roll`, with the probability to succeed being the orange area under the curve, which starts at `x > 1.0`.
![Prof1](https://user-images.githubusercontent.com/76401419/119374795-4f947380-bcba-11eb-96aa-9668741cad33.png)


Now, if a recipe has a 2x failure multiplier, the ratio is divided by 2, which is equivalent to `diff_roll` being multiplied by 2. This leads to the following densities. The orange area is barely visible on the right plot (the area is equivalent to roughly 0.01%):
![Prof2](https://user-images.githubusercontent.com/76401419/119374271-ab123180-bcb9-11eb-97ab-53c6eb322118.png)


This illustrates why the effect of proficiencies seems to be so extreme all of a sudden. I do not think that when the default failure multipliers were set to the levels they are, it was foreseen that the effect would be this large.

The plots were created by replicating the success_roll code in python, then creating 1,000,000 example rolls for each roll type and applying a Gaussian kernel density estimator.

**Comment by KurzedMetal** [comment]
>  merging #48673 fixed one problem, but also made crafting even very simple items practically impossible without proficiencies and left missing proficiencies impossible to mitigate because that's also broken - a worse situation, in my opinion!

Probably merging ANYTHING that improves this situation is better than we currently have right now.

All the builds from "Experimental Build #1" (about ~28 days ago) until now can't be played due to this game-breaking crafting issue.

**Comment by Kelenius** [comment]
@pehamm Thank you! This is what I was trying to explain "The impact the missing proficiencies have is absolutely insane", but this is much clearer.

@KurzedMetal Merging the PR is what started all these problems - previously missing proficiencies increased your success chances instead of decreasing it, e.g. missing principles of leatherworking made your crafting 5 times MORE likely to succeed instead of LESS. When the PR got merged and the bug got fixed - that's where crafting became so difficult because all other bugs came to light - before that they weren't noticeable because the success rate became stupidly high which hid all problems.

**Comment by KurzedMetal** [comment]
I know that, I'm just saying the current state of the code have been like this for 30 days and merging absolutely anything that alleviate the problem is better than leaving it like this more time.

And now that the Launcher has started to deliver the latest GitHub Actions builds, more and more people are being impacted like this and starting to create new threads in Reddit about it.

**Comment by actual-nh** [comment]
Ping: @kevingranade?

**Comment by ennob** [comment]
I feel like this shouldn't be postponed because of feature freeze. This is a game-breaking bug that effectively means it is impossible to craft certain items (makeshift welder and electrohack are recent examples where I got stuck). Can we add this to the blockers/To Do list for 0.F? Alternatively we could drastically reduce the failure chance (or the amount of progress lost) as a temporary fix until a rework can be completed.

**Comment by Evrze** [comment]
The difficulty rolls are normalized to the point on item progression (on the item counter itself), if you cleared up to that point you can still do more work towards item completion, you would just incur a penalty to your success and it will generate more failures along the way as you continue crafting.

This is how crafting any item works, people just interpret it all in a binary way when the difficulty rolls are tied towards the crafted item's degree of progress (item counter).

The difficulty of any crafting item even at full requirements gets increasingly harder to generate success as you progress towards item completion. But there is the illusion where it's out of player expectation that the difficulty is constant and they should be able to complete the item.

Since there's just an increased amount of ticks and multipliers that shifts your success rates and ability to make progress, failure point thresholds are just generated at levels you can not clear because the fail counter is incrementing up with every tick of failures that are being generated. At the same time, triggering them resets the counter and scales up your success chance by a small amount because it could be considered as part of a quota to fill to finish the item itself if you do not have the ability to clear the next segments of progress.

You just have to accept that you can't clear it in a single craft since you're required to do more work to complete the item if it's missing modifiers, but depending on how many you're missing it just becomes impossible to craft.

I just think that the numbers are a bit too higher, and it is out of expectation for people. Entry-level recipes should be scaled down to more reasonable amounts like ".5x more likely to incur failure" but convert that to "50% more likely to incur failure" so they know what to expect.

**Comment by ennob** [comment]
It is not just about expectation. For example in the latest case with the electrohack I consistently get to about 77%, fail, loose some materials, and get knocked back to between 50 and 60%. When I continue I get to around 77% again, only for the same thing to repeat. The problem is that I am stuck between 50 and 77%. I literally tried 15 times, spending more resources each time. If only I would have gotten slightly further each time it would have been OK, but I didn't. 

If the player is not actually able to ever complete the item, then the item should have a higher skill requirement. Saying that "this is how crafting any item works" is not helpful. 

One option to help fix this might be the following: After the player fails and looses some progress (and maybe some materials), make the player immune to failure until they reach the completion level where they previously failed. That way the player could still fail many times but he would at least make forward progress each time he invests more materials. It also makes sense from a story point of view since if the player just tried and failed some part of the crafting process, they are unlikely to make the exact same mistake again.

**Comment by Kelenius** [comment]
Plus, the current crafting system makes it literally impossible to complete the item until you roll more than 1 success rate, because there will ALWAYS be a failure point between the current progress and 100%.

**Comment by pehamm** [comment]
> Plus, the current crafting system makes it literally impossible to complete the item until you roll more than 1 success rate, because there will ALWAYS be a failure point between the current progress and 100%.

I have not seen that, can you show me where this happens in the code?

**Comment by Kelenius** [comment]
> I have not seen that, can you show me where this happens in the code?

`activity_actor.cpp`, `craft_activity_actor::do_turn`, near the end checks if the craft is over a failure point. If it is, it triggers a failure, which calls `craft.handle_craft_failure`, which calls `set_next_failure_point`. And because of how `set_next_failure_point` works, unless `crafting_success_roll` returns at least 1, the next failure point will always be between the current progress and 100%, so you'll hit it again. #48924 is currently attempting to resolve that.

Triggering a failure also always sets you back by some percent, but even if it didn't, you'd still be triggering failures over and over as you approach 100%.

**Comment by pehamm** [comment]
I meant that part:
> because there will ALWAYS be a failure point between the current progress and 100%.

This sounds like you are guaranteed at least one fail. But you can get a `success_roll` above one immediately, in which case you will succeed without any fails.

**Comment by Kelenius** [comment]
Yes, if you roll above 1 at any point, you will finish the craft. If that's the first check, then there will be no failures at all.

But you cannot finish the craft without rolling at least a 1.

**Comment by Evrze** [comment]
It's just part of the requirements, the x2 time just generates a x1 more likely to incur failure or x2 proficiency modifier. The difficulty is the same as the previous craft in you count purely on success rate per tick (but awarded 1/2 progress per tick). As, before the craft is finished you have failed twice as much as before even when you succeed twice as much.

In the end, you have 1 item. You did the work of creating 2 items but got 1. All in all, everything seems consistent, it should be working towards "doing more work" and getting less out of it.

Failure multipliers seem to skew the difficulty at the higher portions towards that direction instead of preserving the frequency of events and both will define and multiply each other differently.

The only way to cancel out the failures is by going over the requirements by the same multiplier for success. 

"x more likely to incur failure" must be canceled out with an equal amount of "x more likely to succeed" since they are both applied for base addition and multiplication. If you standardize success rolls to values 1 and have a proficiency multiplier of 3, you need to calculate for base (3/3) success_roll 1. Which goes over the requirements.

If you are being segmented during the craft and trigger a failure point early, it means you never had the ability to roll for that 1.
Failure is a requirement to finish the item, if you failed early it means your failure point counter already reached it's limit to trigger the failure point.

You compensate the modifiers by doing more work and failing enough to fill the average amount expected before it will let you complete the item. Rolling in each portion of the craft per tick is a form of canceling out the time modifiers at each segment, if it's making you lose progress at a certain point it means you just need more requirements there to cancel out the modifiers.

**Comment by Evrze** [comment]
I forgot to mention, people are usually a degree off when they use the multipliers, x5 failure literally means 500% failure rate on the crafting recipe.

The associated bonus tied to having the proficiency is a time multiplier of (1/2) and a "x1 more likely to succeed".

The proficiency mitigation system and values are consistent, is just that people are off by an order of magnitude that they aren't expecting.

**Comment by Evrze** [comment]
Normalization of a zero value is associated with the event of success and failure. It is simultaneous determining it as a (1) or whole while treating it as a point of referencing other values.

Photons have no mass, but there is negative mass, the high energy collisions of photons create a mass: an electron-positron pair. Photons are a normalization of defining mass to zero value.

The observation of creating success and failures that tie to a value of "negative success" and "negative failure" is occurring due to the normalization of values associated.

All things are consistent and tie into one complete system that requires a normalization of values to create the effect requiring "negative progress" to be made in order to successfully finish the item.

**Comment by samuelmattjohnston** [comment]
Why not revert the "fix" of the formula for now. Fix all these issues in a future version, that way you can not be blocked with so many balance changing question last minute for this release.

**Comment by Evrze** [comment]
It's a feature of the code itself, the amount of time it takes to craft the item is predetermined by its time multiplier, even if it's fluctuating up and down, it's technically doing a linear amount of work. Just treat finishing the task when the time is up, if it isn't up, it isn't ready.

It's based on the normal values that it's being scaled by, how it goes up and down is suppose to provide for the illusion of probability while preserving the frequency of events. The guy who coded it is doing it right. If you fill the failure quota you're done, You should associate the property of a "negative success roll" crafting with failures and it gets translated into being more likely to fail with progress lost in order to be actual success roll. Eventually, you can get a crafted item at the cost of more effort to create.

If you have the proficiencies you actually have the bonus of a (.5) time associated with the normal value with an increased chance to succeed compared to it.

But technically it's also treated as the base outcome for 75% success and 25% failure. Which is normal.

You could just show them the supposed bonuses to having the proficiencies, technically it still is true. You are 1x more likely to succeed than normal.

Mitigation seems to only work towards going to a 0x multiplier towards the normal values. It's why the calculations are so wonky and the recipes as well. Because of how they can be used as but multipliers and for base addition and subtraction for it to work as well as how it's limited to being stored.

If you start 1% of a craft and it doesn't generate the same set of outcomes after if you compare reloads then something is broken.

**Comment by I-am-Erk** [comment]
After some discussion in discord, I think there are two primary changes to make for stable regarding fail rates.

1. Dramatically reduce fail multipliers in proficiencies. I suggest changing them to 1+ (1/10 current value), so a fail mult of 5 would become 1.5. This may be a bit too far, it will require a little testing... at a guess it's goign to lie somewhere between that and 1+(1/5 current) where the highest mult would now translate to 2.
2. Consider having current stage of crafting impact chance to fail, by either adding `current progress/2` or just straight up `current progress` to your crafting roll. The former will preserve some of the current effect of failing more at higher percent completion, which is perhaps more accurate to real life, while the latter should more or less eliminate it entirely.

For the raw problem of proficiency failure rates, 1 should solve almost everything and 2 should cover our bases on the rest. Then #46153, which is closer to complete than I honestly thought it was, will turn this all on its head again after stable.

Of note as it came up a few times: we should avoid any string changes for stable. Math changes to solve the bug are appropriate.
This obviously doesn't touch the UI calculations or the effect of books, I haven't looked at those.

**Comment by pehamm** [comment]
I'll start a few simulation runs.

On 2: wouldn't `current_progress/2` guarantee success after 67%? Or am I misunderstanding your intentions? On the other alternative, I made #48924 for just that.

**Comment by I-am-Erk** [comment]
Since it is only adding your `current_progress/2` to your `craft_roll`, which is counted as a success if >1, it shouldn't guarantee success at any point. It just progressively reduces your fail chance the further along the craft is when you begin. I think some degree of failures clustering later on is OK behaviour.

I am OOTL on what #48924 is doing in the actual code, I would need to take a closer look at that.

**Comment by pehamm** [comment]
No, because we add the result to the current progress in line 1052. The PR makes the progress rolls additive, failing at 51% and rolling another 50% leads to success at 101% (ignoring lost progress).

**Comment by I-am-Erk** [comment]
Aaah, I see how it's changed now.

That is also probably a perfectly fine fix.

**Comment by pehamm** [comment]
If the failure rates with missing proficiencies are adjusted down, this can be put off for pretty long. It still peeves me that my probability of failure depends on when I last failed (right now reaching 99% without having failed is much more likely to finish than reaching 99% after failing at 97%), but if people succeed eventually that is no more than a minor annoyance again.

The current system has some elegance in that the failure rates are geometrically distributed and ingnoring skill gain we can say that the expected number of failures equals the reciprocal of the (computable) failure probability.

The downside is that if we want to train proficiencies, it is better to start a new craft after failing with the current system, as the expected time until failure increases.

**Comment by I-am-Erk** [comment]
Yes, reducing the magnitude of proficiency effect is by far the more important fix, but I would still like that bug to be managed: it is not one that is likely to get fixed by the normal_roll conversion unless that starts expanding outward, so since we know it is there we should do soemthing about it.

**Comment by pehamm** [comment]
>     1. Dramatically reduce fail multipliers in proficiencies. I suggest changing them to 1+ (1/10 current value), so a fail mult of 5 would become 1.5. This may be a bit too far, it will require a little testing... at a guess it's goign to lie somewhere between that and 1+(1/5 current) where the highest mult would now translate to 2.

Looking good. I simulated crafting an anvil because it is imo the most annoying recipe as of now due to the recipe needing both Principles of Metalworking and Blacksmithing proficiencies to craft and both proficiencies being annoying to practice without one (I am drowning in steel frames as one of the few viable options for blacksmithing).

Anyway, the recipe needs 3 fabrication, principles of metalworking has a `5 x fail_mulitplier`, and blackmithing a `2.5 x fail_multiplier`. With the old system this would have been a `5*2.5 = 12.5` divisor on the success roll, now it would only be 2.25. With average intelligence and 3 fabrication, we get a success probability of rought 0.05% and the following densities (again, probability densities of both rolls left, density of the ratio right, probability to succeed equal to the orange area under the density starting at `x >= 1`):
![3 Skill](https://user-images.githubusercontent.com/76401419/119909888-44f60a80-bf56-11eb-9c89-c4bef5c9cd91.png)

with 6 skill, we make massive progress getting 28% successes:
![6 Skill](https://user-images.githubusercontent.com/76401419/119909959-62c36f80-bf56-11eb-8daa-92c8701156cb.png)

and with 9 skill, this grows to 95%:
![9 skill](https://user-images.githubusercontent.com/76401419/119910040-8981a600-bf56-11eb-97d3-4113070ed440.png)

For comparison, with the old system I did not roll a single success in a million tries even with 9 intelligence.
![9 skill old](https://user-images.githubusercontent.com/76401419/119910343-3f4cf480-bf57-11eb-9844-0fe355a6c5e6.png)

**Comment by I-am-Erk** [comment]
thank you very much for that @pehamm, that's exactly the sort of data I was looking for and a good choice of recipe. that suggests that if anything the 1+1/10 model might still be too tough, but it is definitely on a reasonable scale now.

This is an easy fix. I will set up a PR.

**Comment by Evrze** [comment]
Part of the reason you guys are failing more is because of how the proficiency multiplier was being used for multiplication is now used for division.

Basically, if you were playing on updates before the change, the failures were spread out much more evenly and your fail counter just incremented upwards by the same value. You just compensated with time and your failures were spread out much more evenly.

Now for division, it's treated as an opposite effect one failure is multiplied by "x more likely to incur failure" under the way it's now scaled.

Both directions work with the current code, they just scale differently to achieve the same thing, you see it as a high failure rate because you suppose to see them in bigger bursts. You have a very small failure counter in comparison to the multiplier affecting it.

I don't really think there's anything wrong with the way it is now, overall as long as the frequency of failures is kept within the expected range it's the same to me.

**Comment by Zireael07** [comment]
Whatever ends up done, the sheer amount of discussions and misinformation about this problem show that this part of the code needs to be well documented/commented to avoid getting broken sometime in the future again...

**Comment by Evrze** [comment]
Well technically time and failure rates have inversive properties tied to the recipe modifier by base manipulation, the solution and problem itself is tied together.

If a solution of compacting the time multiplier and ends up increasing the failure rate, you see it as a problem but it's tied to how the code works with the proficiency system to calculate everything.

If you want to increase the recipe's "time multiplier" to a point where you never fail, technically that can happen but that's going to create more problems.

**Comment by I-am-Erk** [comment]
Evrze, I don't mean to be harsh here, but I'd really appreciate if you'd back away from this topic a little. As best I can tell you don't have a really good grasp on it and your enormous posts have actively slowed me down from fixing it dramatically as I try to figure out what you're saying and how it relates to anything. I know you mean well, but it's clear you don't have a deep understanding of the issue; this is getting in the way.

**Comment by Evrze** [comment]
Alright, I've said all I need to say.

**Comment by Evi-d** [comment]
A high failure looks necessary since inversion of the proficiency multiplier shifts the time compensation from duration into failure chance.

In the end, it's just rolling for success and failures, if you have any failures the craft can not be complete because it looks like it's recalculating for a success roll 1 for a progress bar roll 1. It should make it impossible to clear with doing a failure roll which could be considered a feature, after clearing out your failure rolls it lets you progress forward.

Players are just seeing their failures build-up towards the end because the progress bar on the display can only show so much, it has to clear out all your failure rolls from the previous step in order to build up because your success die can fully cover all of the areas involded.

It looks like it's a nerf to your skills but if it compresses your success die, since it's the inversion of the prof_multiplier it creates an effect where your die is considered to have fewer failures but requirements more failure rolls to complete the craft.

If it's working correctly it should just be making crafting times shorter but the time multiplier isn't really time.

It doesn't really need fixing, players should just adapt to it as a new type of crafting method calculation.

**Comment by Evi-d** [comment]
Do people not understand the number of rolls that are required to fill a progress bar on the display up?

Literally, you're rolling real-time per tick while it fills up as time is passing, the change does reduce the number of rolls it takes so I see it as fine. If you all are treating it in a binary in two rolls you won't see the change, it's not binary, it was never binary even before this.

If you have a success roll (.90) and roll 10 times you get .35.

What happens to the other part where your success roll isn't 1? It's a failure roll, then it's picked from one of the failure outcomes.

Increasing failure rates and reducing the number of rolls has a similar effect but just generates a higher perceived failure rate since you'll see them more frequently.

Since it seems like no one understands probability, math, and coding the changes we make break the game state even more even if it's originally correct.

I legitimately am confused, fixing things to make them broken seems to always be the trend, plugging around numbers that you don't know what they're tied to just create that effect.

If you had to do 400 rolls but now you only do 200 rolls but at a higher failure rate isn't it fine?

It can achieve a similar effect even if it's more punishing, there is always a correlation between the number of rolls and failure rates because a success roll (1) can never always be a success roll (1) for every area as it continues rolling.

You have to trigger a failure outcome at some point and choose from one of the effects.

This has been the case even before the change, you were just incrementing it and spreading it out in a larger number of rolls, that is just what the "time multiplier" affects, it's the number of rolls, it's not the same "time" as the game ties it to but a standard of rolls per craft.

People who do not understand what the numbers are tied to will not be able to accept the change. If you want to do more rolling and reduce your failure chance that's fine too, reduce the number of rolls.

All I'm saying is that it does have the benefit of cutting back on "rolls" per craft, if even a tick of your progress bar even by an increment is a simulation of a roll. Why are people trying to compress it into one or two rolls?

It is the definition of insanity if you try to do that, that will never work because you end up inflating your rolls to be something it isn't. You can sit and wait for your roll 1d8 to roll 100 and it will never work. What you clear is what you work towards per tick, why can't people see that?

You want to split it into two rolls and it will never work, do you roll once for 50% progress on the display and 50-100% progress on the other?

That is never how it worked and never how it should work.

**Comment by Evi-d** [comment]
I have been emphasizing the differences between success chance and success rate as well as failure chance and failure rates.

People tend to get them confused with each other but they are completely different.

This is why a 1% success chance can have consecutive successes in 100 rolls, and you can have a success rate that offsets it, but in the end, you need more rolls to finish the item, what if there is a failure rate modifier that offsets your success chance and increasing your failure chances?

If you're sampling for 100 rolls, you require 100000 rolls to complete, what makes anyone think that your success roll has an equivalent effect towards every portion?

Your progress bar always had failures in them, but the way they incremented before made you not care too much about them, now since everything much more compressed, you have to see the failures go up since the rates have to be different to scale it down, otherwise, it wouldn't express the same thing.

If you scale down to 100 rolls, are you going to expect the game and code to give you a 100% success chance with a success roll (1) if you had a success chance of 1% at the start?

Of course not, you can tweak how it scales, but in the end, it's going to affect it both ways. Scaling down the number of rolls will always increase the failure rates in order to modify the situation towards the expected results.

There are no particular benefits with what we have now, but at least you can see the portions you have completely cleared out, if the game pushes your progress back, you can always climb to higher completeness afterward.

**Comment by Evi-d** [comment]
https://medium.com/@tglaiel/how-to-roll-half-a-dice-21631c1fe694

This is the effect of cutting the rolls in order to represent the new values, what you had before was a different "shape" than what we have now.

The normalization factors were just holding you accountable towards a certain factor away from it. A x.1 more likely to incur failure is roughly less than 1.1 from the normal point.

Please understand the values of your die, these are just going to be the effects of what it is to hold you accountable for every outcome that occurs.

Your success is built from your failure and your failures are built from successes, it must be forced to be that way. You can change the "shape" of the distribution if you want but I'm saying how it works now is absolutely fine and functional but require a different set of interpretation.

There -.x time multiplier was making the base value from the normal point essentially lower than what it is, you need that "negative time" for cutting down rolls from the normal point. Now since it's the division you must need to have a higher failure rate when you cut down the "time".

As you continue, your success rate must continually decrease to represent:

"function: roll one continuous dice, apply f(x) = x^(1/n)"

You can treat everything that is happening in your progress bar as the effects of that continuous die constantly being rolled.

Progression rollback is necessary by definition if you cannot fulfill the completion of the item before that.

It's not intentionally there to punish you, it's just the shape of what you're rolling for must be the accepted new outcome of events, technically you will always fail more at the end, if you remove the item counter it will stop spawning unclearable failure points and just let you do segments normally, but then you no only preserve the same probability of events in the frequency that must occur. The failure points mean you're filled up your success quota number and not allowed to have any more success rolls and now must pick from a failure outcome if you want to continue, after picking a failure outcome you accept the outcome and continue rolling for successes.

The shape of "multiplication" of the prof_multipler and the shape of "division" of the prof_multiplier is already different from how things are rolled for.

If you want to distribute the failures more evenly it can only be done through the multiplication of the prof_multiplier and reverting back to how things were before. 

Overall, if you already adapted to the new "shape" of the failure distribution by reducing the number of die rolls you can just keep the change and consider it as a feature.

**Comment by Evi-d** [comment]
Die rolls cannot be simulated in normal distributions because they should always be discrete distributions.

https://corporatefinanceinstitute.com/resources/knowledge/other/discrete-distribution/#:~:text=A%20discrete%20distribution%20is%20a,%2C%2010%2C%2015%2C%20etc.

It must store the value of your failure somewhere and force the failure outcome, if you have a die at some point it will fail, the value of that failure must be stored.

It seems like no one understands what the code does even when they look at it, so when it's tweaked, no one understands the effects of what is supposed to happen.

A binary interpretation of results can never exist in a continuous die that simply rolls until the end of item completion, that is what you see on the item progress bar (display) from 0 to 100%.

NEVER use a binary interpretation of this to simulate the chain of events.

It can never be a normal distribution because it takes place in discrete mathematics, if you force it to be a normal distribution it simply will not ever work as you expect it to be.

**Comment by Evi-d** [comment]
For anyone who can actually do the calculations and want to, the bottom of the recipe is for a lognormal distribution from the normal point you should be calculated from. There could be negative values as well that can modify it or positive ones to represent direction from either side.

Apparently, most people don't know what the "negative time" multiplier is for. It's for the other side of this distribution. They go both ways.

It is not a normal distribution because it has the elements of skewness involved.

It was "x2 long as normal" and "x1 more likely to incur failure" and you can see this as binary logarithm, or log base 2.

But it's from the normal value.

Skewness has to be represented, which can only occur in lognormal distributions. If you're ever using a normal distribution it's not going to work.

In general, the code can is seen as a calculator that outputs a function that seems consistent enough. I don't know what you expect.

**Comment by Evi-d** [comment]
If people haven't realized how inverted how the prof_multipliers affect the skewness of gearing everything towards failure I don't know what to say.

Technically it's just making everything harder with no benefit for the players except that stop them, which is great in some ways if that's what you're looking for.

The "right answer" that was given to me enforces failures, so I don't know what to expect, when the rates already showed how skewed the distributions are I'm not sure why people are still trying to do skill rolls like it was before.

There have never been any right answers though and only answers people are willing to accept as right.

The guy who took the inverse of the prof_multiplier to multiply to the skill roll had no clue what he was doing or what he was affecting.

The distributions are fucked up because of that to push everything to failures rates instead of compensation with "time" like it was before. 

Now you are forced to do a "failure compensation" instead of a "time compensation", it still works the same way but with a different compensation system.

I don't think it's balanced, but at the same time it does cut down on the rolls you do, he did intend to make everything harder. It does have that effect.

All in all the end result is achieved even though he wasn't sure what he was doing. No one is sure what they're doing apparently.

Until you do your "failure compensation" with more rolls you should not be allowed to finish crafting currently, which is what he seems to have wanted to do....

Why do you see it as a problem when you set it up to work that way?

I guess the work people do is the problem. Then they don't see a problem in their work, I suppose we can all treat this as a learning experience.

I'm ok with the failure compensation system and people have been playing on it for a while, yes the skewness leans towards failures because technically the outcome of a normal event is supposed to be set to:

75% success and 25% failure 

You start your rolls at 100% and then end on 25% towards the end of the craft due to the skewness that has to be expressed. This was how the distribution always occurred just due to the crafting system itself. 

What the guy had no clue about is the rates also define each other, .25 failure chance is 1/3 of .75 success chance, success chance is 3x failure chance.

The log distributions are fucked now because of it, technically it could just be fine to accept it as is, the item counter will always make you fail more than you need by spawning impossible to clear failure points to meet with its expectations.

This sets a very high difficulty settings for players at the end and in the beginning especially before they have proficiencies. The normal event of occurrence are always failures like the guy intended to lean things towards. Your failure point counter now has a multiplier to help you reach your failure rolls faster between 0-100% of the craft especially at the last parts.

**Comment by Evi-d** [comment]
If you go to the casino and play the slots, you can expect to always lose more money than you put in.

The house always wins, it's rigged that way.

What happens if you win a million-dollar jackpot?

Even if the machine says you do, people are going to stop you. They'll pass it off as the machine malfunctioning and give you $100 dollars and a free meal at the buffet.

You can rig your own slot machine to give you more money than you put in, that works too if it feels satisfying for you. In the end, the money in the machine is yours to play with. If someone else uses it naturally you'll just be burning your own money.

You can do a virtual slot machine and keeps giving your virtual money and success and have it print out as much as you like. I'm just saying if they skew the rates too much in your favor you don't feel any effects at all. The skewness of the failure rates due to the "faster" rolls you have to make now just sets you up for it.

I'm just being the person that says, oh if you actually succeeding at the end then the slot machine is malfunctioning or you cheated the system.

Just flip the switch and put it back to being a multiplier and don't invert it as a multiplier.

There is a huge different between x and (x)^-1.

This is a base operation of multiplication and division.

Since the numbers are lumped together like that you can do anything to them.

It's called a multiplier for a reason but the guy took the inverse of it, or essentially divided it and multiplied it to the skill roll.

If we go by casino rules, it's going by the house always wins. You must fail, but at least the game has the decency to show your failure rates, just fail in the end x5 times the normal rates where you expect it to be a 50-50 chance at the normal rates.

Since there are so many contradictory values that a failure rate and success rate can influence it to be a 50-50 chance there should be impossible to do a skill roll the ways you guys have been trying to do things.

Even the people that are coding it don't even know what it does apparently adding more confusion to the mix that's being continually passed along to everyone at every section.

Everyone is trying to fix it in their right way creating a very interesting lump of problems. We can just let it go back to being a multiplier that isn't multiplicative inverse. 

Everyone is actually falling for the "gambler's fallacy" if they think they can pass what they cannot pass. You'll always get stopped somewhere along the way. Technically speaking, there was never any "success crafting", so you guys have always been using the wrong skill roll method to see the actual outcome of events.

Even before the change you failed a fuck ton with messages that "you lose progress" that meant the tick didn't succeed into adding progress but your skills mitigated the effect. Someone removed it and the system shifted towards what we have now, it's actually hard to keep track of your failures now but I can assure you, you're failing a fuck ton more if you're segmented, you just don't see how it rolled. Multiple failures at higher difficulties are assured for progression lost and additional material with item destruction because you cannot mitigate the failure outcome penalties at the end.

If you don't see how it's rolled how can you even put it into a simulation or graph or interpret it?

You literally failed a fuck ton and ignored it all in the beginning segments, the last segments have the highest difficult and you expect to succeed? 

You are literally ticking down your success when the rates are skewed by definition.

People are trying to say to me that "that literally not how it works" just makes them seem retarded to me. They don't even understand what I'm trying to say, I'm saying the average failures have to be the same. They are segmented as an expectation, "consecutive success" is translated into failure rates to affect the failure chance of the proceeding rolls.

How else do you think the slot machines are rigged. Every time you succeed it's rigged to make you fail more, the guy who coded the program initially just set you up for success, now it's set for failure.

We can stop going by casino rules if we revert, to not inversing the multiplier as the easiest fix or we adjust the casino rules. Just know that it works in both directions if you flip the switch.

It simply seems like no one has the right interpretation of what is supposed to occur and enforces their own. Anyone who knows what a "failure quota" implied should have linked it all together. 

Why do you think the probability is preserver between each roll with each being independent of the other? They literally have to be linked together in order for probability to even be expressed. They mutually exclusive and must be, but at the same time affected by any rates and modifiers that can skew them towards both directions.

If one is triggered the other isn't, why the fuck are some people trying to do it all in one roll and treating it as a binary event?

Since the guy implied the original coder of the proficiency system doesn't know what he's doing, I'm just saying I have to do it to everyone around here as well, especially for people who look at the code but at the same time don't seem to understand rates and probabilities. 

Fix your delusions to how things really work, please. Do not inflate what your rolls are and should be doing. The probability exists because it is enforced, if you don't know the rules, you don't know the rates, you don't know the settings, you don't know anything.

In general, that's why it's best to go by frequency and work back from there because can have expectations of how it is enforced and work backward. But it will never actually give you an absolutely correct answer, just an expectation, because you don't know what it really is calculating it all based on.

If you want to jam in more numbers into the machine for it to work technically it should just already be set to do that, and jam in more until it works in the way you think it should work.

See the solution?

See the problem?

It's the same.

**Comment by Evi-d** [comment]
What people are trying to treat as their skill roll from both players and programmers are actually "negative skill rolls" right now.

You set your expectations to be able to do it, but it should never occur.

You know each die has a set mathematical pattern that it expresses and cannot be broken as that is an enforcement of the rules as well. You use the wrong die, you use the wrong rules, you rolls have broken the rules of what should be allowed. That die isn't going to give you the same results at some point.

To compensate, you use frequency of events of counter failures, consecutive fails are rewarded with the ability to succeed. It's simply how the rates are skewed that the ability to succeed seems less. High-frequency failures let you have your success rolls faster because they're treated as completed.

The rates are allowed to regenerate or deplete depending on how the rates were initially skewed, rates always affect predetermined numbers that affect the current failure and success chances.

Usually, it should be set to a rate x1 for success chance the same as a rate x1 for failure chance from the normal 50-50 value. Since the "time multiplier" was initially for frequency, but now it inverted, it means the frequency of failures required, naturally you just end up with high-speed failure rolls.

It should be impossible to do a skill roll, that is the illusion it should be giving players and programmers. So it seems to all be working properly but at the same time creates so much confusion on how it should be working. It's already the improper intention of the original code to run towards a failure rate skew. It was keeping a success rate skew by spreading it out before. The frequency was kept the same, but compensated by your success rolls. Technically it did trigger more failures, but it was all ok since your successes were also 1x, the incurred penalties weren't as massive.

Since someone didn't feel the effect was greater enough and rigged it up to go by casino rules, there must be barriers to what you can ever do with your rolls at a higher difficulty which should spawn directly on the tick you triggered it on the item's progression, then you get rolled back and the counter starts ticking again. 

Seems reasonable to me.

The setbacks are unavoidable with little rewards after you succeed, in the end, you always lose more.

That's if you want to go by casino rules though. Technically no rules actually exist if you can change them so might as well, might as well change the rule sets.

In the end, if everyone enforces their own rule sets it's hard to keep track of, naturally experimental versions are always expected to be broken, even stable is broken.

It isn't broken if you shift a variable or multiplier and it affects everything that ties to it. The current state of the game has probably the easiest fix. All that had to be done is use the original log distribution of success and failure rates rather than going by casino rules "downward" towards failure.

Failures are literally the only option for item completion to allow for a smaller degree of success.

Let players have an ability to see themselves do a "skill roll" and succeed rather than a "negative skill roll". They're just spreading out their failures in the end in different ways or compressing them into high-frequency failure rolls.

You should be able to calculate your real probability of passing with a "negative skill roll" with how the current system works. Any positive value for a skill roll does not and will not exist if you want to do probability.

Look at the distribution on the other side, that's the "negative skill roll" that you guys keep ignoring. Look at the other fucking side of the distribution.

That's what you have to work against.,

**Comment by Mom-Bun** [comment]
Locked for now as this issue seems to be getting spammed by one user in particular.

**Comment by Raikiri** [comment]
The experimental has lived for almost half a year with failure chances doing practically nothing and it worked just alright. Isnt it safer to release 0.F with failure chances just completely disabled until the issue is addressed properly?

The reasoning is that the current proficiency system still makes crafting quite hard (as in, way harder than it used to be) even from time penalties alone.

**Comment by I-am-Erk** [comment]
There's no need to disable them, they're working pretty well at this point. Only the book fix needs to be merged now.

**Comment by Raikiri** [comment]
I understand that it's probably fixed now, but if during a feature freeze time, certain bug fixes can have dramatic impact on the gameplay (such as when it turns out certain stats had no effect in practice), then fixing them would produce more work (because now somebody needs to actually make those stats work). It's just natural to leave those types of can-of-worm bugs until after a milestone is reached.

But it's just me hoping the next stable will be reached more smoothly.

**Comment by I-am-Erk** [comment]
It was a bit of a debatable choice to fix the failure rate bug in the first place rather than just disabling it, but despite the kerfuffle over it, it was not in fact a big deal at all to smooth it back out.

**Comment by I-am-Erk** [comment]
I believe this is ready to close, any objectors?

**Comment by Kelenius** [comment]
I agree, the big issues have been fixed.
